### PR TITLE
SIGINT-1650: PRComment is set but no SCM resources are set

### PR DESCRIPTION
**Pull request by kiransynopsys**

### CODE REVIEW

**Review by sig-tithi — APPROVED** [review]
Looks good to me. but also get it reviewed by @spurohit and @kishorikumar

**Review by sig-tithi — COMMENTED** [review]
return undefined; can we have a better error handling?

File: synopsys-task/src/synopsys-task/tools-parameter.ts [hunk 1]
```diff
@@ -445,6 +445,18 @@ export class SynopsysToolsParameter {
       const parsedUrl = url.parse(collectionUri);
       azureInstanceUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
       azureOrganization = parsedUrl.pathname?.split("/")[1] || "";
+      if (!azureOrganization) {
+        const protocolSeparator = "://";
+        const protocolIndex =
+          collectionUri.indexOf(protocolSeparator) + protocolSeparator.length;
+        const domainIndex = collectionUri.indexOf(".visualstudio.com");
+        if (protocolIndex >= 0 && domainIndex >= 0) {
```
Comment (kishorikumar) [inline synopsys-task/src/synopsys-task/tools-parameter.ts:453]: I think, if it is a valid uri then index cannot be 0.

File: synopsys-task/src/synopsys-task/tools-parameter.ts [hunk 2]
```diff
@@ -445,6 +445,18 @@ export class SynopsysToolsParameter {
       const parsedUrl = url.parse(collectionUri);
       azureInstanceUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
       azureOrganization = parsedUrl.pathname?.split("/")[1] || "";
+      if (!azureOrganization) {
+        const protocolSeparator = "://";
+        const protocolIndex =
+          collectionUri.indexOf(protocolSeparator) + protocolSeparator.length;
+        const domainIndex = collectionUri.indexOf(".visualstudio.com");
```
Comment (spurohitsynopsys) [inline synopsys-task/src/synopsys-task/tools-parameter.ts:452]: can we gather org name from hostname itself, we can split the hostname by dot(.) and get the info

Comment (kiransynopsys) [reply synopsys-task/src/synopsys-task/tools-parameter.ts]: seems to be good idea, will fetch from hostname.

File: synopsys-task/src/synopsys-task/tools-parameter.ts (hunk 1 above)
Comment (kiransynopsys) [reply synopsys-task/src/synopsys-task/tools-parameter.ts]: will fetch from hostname.

**Comment by kiransynopsys** [comment]
pieplinejobs: https://synopsysorg.visualstudio.com/NodeGOAT/_build/results?buildId=1792&view=results

**Comment by kiransynopsys** [comment]
> return undefined; can we have a better error handling?

When Azure data is unavailable in the Blackduck/Coverity/Polaris interface, we need to exclude azuredata  which is optional while preparing input command. 
const inputJson = JSON.stringify(blackduckData);
Thus, to resolve this issue, it's important to ensure that azureData is specifically defined as undefined.

Additionally, Added debug message when azure data is undefined

File: synopsys-task/src/synopsys-task/tools-parameter.ts [hunk 3]
```diff
@@ -445,6 +445,13 @@ export class SynopsysToolsParameter {
       const parsedUrl = url.parse(collectionUri);
       azureInstanceUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
       azureOrganization = parsedUrl.pathname?.split("/")[1] || "";
+      if (
+        parsedUrl.host &&
+        !azureOrganization &&
+        parsedUrl.host.indexOf("visualstudio.com") !== -1
+      ) {
+        azureOrganization = parsedUrl.host.split(".")[0];
```
Comment (spurohitsynopsys) [inline synopsys-task/src/synopsys-task/tools-parameter.ts:453]: add only if parsedUrl.host.split(".")[0] exist

File: synopsys-task/src/synopsys-task/tools-parameter.ts [hunk 4]
```diff
@@ -464,6 +471,11 @@ export class SynopsysToolsParameter {
         "Missing required azure token for fix pull request/automation comment"
       );
     }
+    console.debug("Azure Instance URL:", azureInstanceUrl);
+    console.debug("Azure Organization:", azureOrganization);
+    console.debug("Azure Project:", azureProject);
+    console.debug("Azure Repo:", azureRepo);
+    console.debug("Azure Repo Branch Name:", azureRepoBranchName);
```
Comment (spurohitsynopsys) [inline synopsys-task/src/synopsys-task/tools-parameter.ts:478]: use taskLib.debug

File: synopsys-task/src/synopsys-task/tools-parameter.ts [hunk 5]
```diff
@@ -498,7 +510,7 @@ export class SynopsysToolsParameter {
       }
       return azureData;
     }
-
+    console.debug("Azure data is undefined.");
```
Comment (spurohitsynopsys) [inline synopsys-task/src/synopsys-task/tools-parameter.ts:513]: use taskLib.debug

File: synopsys-task/src/synopsys-task/tools-parameter.ts (hunk 3 above)
Comment (kiransynopsys) [reply synopsys-task/src/synopsys-task/tools-parameter.ts:453]: this would be sufficient to set the orgname

File: synopsys-task/src/synopsys-task/tools-parameter.ts (hunk 5 above)
Comment (kiransynopsys) [reply synopsys-task/src/synopsys-task/tools-parameter.ts:513]: using tasklib lib

File: synopsys-task/src/synopsys-task/tools-parameter.ts (hunk 4 above)
Comment (kiransynopsys) [reply synopsys-task/src/synopsys-task/tools-parameter.ts:478]: using tasklib lib

**Review by spurohitsynopsys — APPROVED** [review]

**Comment by officeSpacePirate** [comment]
Is there anything holding this up from being merged/released? This appears to be blocking for the PR comment feature.

File: synopsys-task/src/synopsys-task/azure-service-client.ts [hunk 6]
```diff
@@ -19,14 +19,17 @@ export class SynopsysAzureService {
...
+          (match, index) => encodeURIComponent(args[index]) || ""
+        );
 
       const endpoint = StringFormat(
         azureData.api.url.concat(this.azureGetMergeRequestsAPI),
         azureData.organization.name,
         azureData.project.name,
         azureData.repository.name,
-        "refs/heads/".concat(azureData.repository.branch.name),
+        azureData.repository.branch.name,
```
Comment (sig-tithi) [inline synopsys-task/src/synopsys-task/azure-service-client.ts:32]: azureData.repository.branch.name will already have refs/heads/ appended?

Comment (spurohitsynopsys) [reply synopsys-task/src/synopsys-task/azure-service-client.ts:32]: yes, Both in PR and Non PR Context

File: synopsys-task/src/synopsys-task/tools-parameter.ts [hunk 7]
```diff
@@ -437,14 +437,31 @@ export class SynopsysToolsParameter {
...
     const azureProject =
       taskLib.getVariable(AZURE_ENVIRONMENT_VARIABLES.AZURE_PROJECT) || "";
     const azureRepo =
       taskLib.getVariable(AZURE_ENVIRONMENT_VARIABLES.AZURE_REPOSITORY) || "";
+    const buildReason =
+      taskLib.getVariable(AZURE_ENVIRONMENT_VARIABLES.AZURE_BUILD_REASON) || "";
     const azureRepoBranchName =
-      taskLib.getVariable(AZURE_ENVIRONMENT_VARIABLES.AZURE_SOURCE_BRANCH) ||
-      "";
+      buildReason == AZURE_BUILD_REASON.PULL_REQUEST
```
Comment (sig-tithi) [inline synopsys-task/src/synopsys-task/tools-parameter.ts:458]: is it applicable for both classic editor and yml?

Comment (spurohitsynopsys) [reply synopsys-task/src/synopsys-task/tools-parameter.ts:458]: Classic editor is not considered as pull request event, branch name will be fetched from Build.SourceBranch only.

File: synopsys-task/src/synopsys-task/azure-service-client.ts (hunk 6 above)
Comment (sig-tithi) [reply synopsys-task/src/synopsys-task/azure-service-client.ts:32]: can we add some test cases to handle negative scenarios to make sure if something else come it has proper error handling.

Add debug logs for env variables or data that is not set directly by user but derived internally.

Comment (spurohitsynopsys) [reply synopsys-task/src/synopsys-task/azure-service-client.ts:32]: updated test cases accordingly
